spec/haskell/src/SEL4/Object/SchedContext.lhs

```diff
@@ -297,11 +309,42 @@ This module uses the C preprocessor to select a target architecture.
 >     setRefillHd scPtr (Refill { rTime = rTime (refillHd sc), rAmount = rAmount (refillHd sc) + rAmount (refillTl sc)})
 >     setRefillTl scPtr (Refill { rTime = rTime (refillTl sc), rAmount = 0})
 
-> refillHdInsufficient :: PPtr SchedContext -> Kernel Bool
+> refillHdInsufficient :: PPtr SchedContext -> KernelR Bool
 > refillHdInsufficient scPtr = do
->     sc <- getSchedContext scPtr
+>     sc <- readSchedContext scPtr
 >     return $ rAmount (refillHd sc) < minBudget
 
+> nonOverlappingMergeRefills :: PPtr SchedContext -> Kernel ()
+> nonOverlappingMergeRefills scPtr = do
+>     old_head <- refillPopHead scPtr
+>     updateRefillHd scPtr $ \head -> head { rTime = rTime head - rAmount old_head,
+>                                            rAmount = rAmount head + rAmount old_head }
+
+> headInsufficientLoop :: PPtr SchedContext -> Kernel ()
+> headInsufficientLoop scPtr = whileLoop (const (fromJust . runReaderT (refillHdInsufficient scPtr))) (const (nonOverlappingMergeRefills scPtr)) ()
+
+> mergeRefills :: PPtr SchedContext -> Kernel ()
+> mergeRefills scPtr = do
+>     old_head <- refillPopHead scPtr
+>     updateRefillHd scPtr $ \head -> head { rTime = rAmount old_head,
+>                                            rAmount = rAmount head + rAmount old_head }
```

Comment:
I might be misunderstanding things, but this `mergeRefills` is used in `refillUnblockCheck` and doesn't seem to match what the C does (which is `refill_head(sc)->rTime = NODE_STATE_ON_CORE(ksCurTime, sc->scCore)`).

Reply:
Curtis is happy to change this to the way I have it in this PR. See

https://github.com/seL4/seL4/pull/348#issuecomment-826503214

Reply:
Awesome, that will help crefine!